nnet_utils/nnet_activation.h

```diff
@@ -325,6 +320,316 @@ void  tanh(data_T data[CONFIG_T::n_in], res_T res[CONFIG_T::n_in])
+    int index;
+    for (int ii=0; ii<CONFIG_T::n_in; ii++) {
+        if (CONFIG_T::io_type == io_serial){
+            #pragma HLS PIPELINE
+        }
+        data_round = data[ii]*CONFIG_T::table_size/16;
+        index = data_round + 8*CONFIG_T::table_size/16;
+        if (index < 0)   index = 0;
+        if (index > CONFIG_T::table_size-1) index = CONFIG_T::table_size-1;
+        res[ii] = (res_T) softplus_table[index];
+    }
+}
+
+// *************************************************
+//       Softsign Activation
+// *************************************************
+template<class data_T, class res_T, typename CONFIG_T>
+void  softsign(data_T data[CONFIG_T::n_in], res_T res[CONFIG_T::n_in])
+{
+    if (CONFIG_T::io_type == io_parallel){
+        #pragma HLS PIPELINE
+    }
+
+    data_T datareg;
+    for (int ii=0; ii<CONFIG_T::n_in; ii++) {
+        if (CONFIG_T::io_type == io_serial){
+            #pragma HLS PIPELINE
+        }
+        datareg = data[ii];
+        if (datareg < 0) res[ii] = datareg / (1 - datareg);
+        else res[ii] = datareg / (1 + datareg);
+    }
+}
+
+// *************************************************
+//       ELU Activation
+// *************************************************
+inline float elu_fcn_float(float input) {
+    return exp(input) - 1.;
+}
+
+template<typename CONFIG_T, int N_TABLE>
+void init_elu_table(typename CONFIG_T::table_t table_out[N_TABLE])
+{
+    // Default ELU function:
+    //   result = alpha * (e^(x) - 1)
+    for (int ii = 0; ii < N_TABLE; ii++) {
+        // First, convert from table index to X-value (signed 8-bit, range -8 to 0)
+        float in_val = -8.0*ii/float(N_TABLE);
+        // Next, compute lookup table function
+        typename CONFIG_T::table_t real_val = elu_fcn_float(in_val);
+        //std::cout << "Lookup table In Value: " << in_val << " Result: " << real_val << std::endl;
+        table_out[ii] = real_val;
+    }
+}
+
+template<class data_T, class res_T, typename CONFIG_T>
+void  elu(data_T data[CONFIG_T::n_in], res_T res[CONFIG_T::n_in])
+{
+    // Initialize the lookup table
+#ifdef __HLS_SYN__
+    bool initialized = false;
+    typename CONFIG_T::table_t elu_table[CONFIG_T::table_size];
+#else
+    static bool initialized = false;
+    static typename CONFIG_T::table_t elu_table[CONFIG_T::table_size];
+#endif
+    if (!initialized) {
+        init_elu_table<CONFIG_T, CONFIG_T::table_size>(elu_table);
+        initialized = true;
+    }
+
+    if (CONFIG_T::io_type == io_parallel){
+        #pragma HLS PIPELINE
+    }
+
+    data_T datareg;
+    // Index into the lookup table based on data
+    int index;
+    for (int ii=0; ii<CONFIG_T::n_in; ii++) {
+        if (CONFIG_T::io_type == io_serial){
+            #pragma HLS PIPELINE
+        }
+        datareg = data[ii];
+        if (datareg > 0) {
+            res[ii] = datareg;
+        } else {
+            index = datareg*CONFIG_T::table_size/8;
+            if (index < 0) index = 0;
```

Comment:
Just tested elu and found that we will always pass this if statement because the indices go from -8 to 0.  Can change this section or the indexing to fix the sign problem.